plugin/plugin.go

```diff
@@ -57,58 +59,66 @@ type CLIPlugin struct {
 }
 
 // NewCLIPlugin validate the metadata of the plugin and return a *CLIPlugin.
-func NewCLIPlugin(ctx context.Context, name, path string) (*CLIPlugin, error) {
+func NewCLIPlugin(ctx context.Context, name, path string, pluginConfig map[string]string) (*CLIPlugin, error) {
 	plugin := CLIPlugin{
 		name: name,
 		path: path,
 	}
 
 	// validate metadata
-	metadata, err := plugin.GetMetadata(ctx, &proto.GetMetadataRequest{})
+	metadata, err := plugin.GetMetadata(ctx, &proto.GetMetadataRequest{PluginConfig: pluginConfig})
 	if err != nil {
 		return nil, fmt.Errorf("failed to fetch metadata: %w", err)
 	}
 	if metadata.Name != name {
 		return nil, fmt.Errorf("executable name must be %q instead of %q", binName(metadata.Name), filepath.Base(path))
 	}
```

Comment:
Should this check be in `plugin.GetMetadata()`?

Next question is "Do we really need to validate the plugin when calling `NewCLIPlugin()`?".

Reply:
1. We need to validate the plugin name of reponse metadata that should be equal with user provided name. In `plugin.GetMetadata()` we don't have the user provided plugin name. 
2. We need to check the plugin name. When we return a new plugin, it should be a valid plugin or return an error. If we don't do the checking here, user also need to check it in another place.

Reply:
> We need to validate the plugin name of reponse metadata that should be equal with user provided name. In plugin.GetMetadata() we don't have the user provided plugin name.

We do have the plugin name as the `name` parameter in `func NewCLIPlugin(ctx context.Context, name, path string, pluginConfig map[string]string) (*CLIPlugin, error)`

> We need to check the plugin name. When we return a new plugin, it should be a valid plugin or return an error. If we don't do the checking here, user also need to check it in another place.

Users always call `plugin.GetMetadata()` to determine the capability before use, don't they?

Reply:
Updated.